cmd/catalyst-node/catalyst-node.go

```diff
@@ -332,166 +135,86 @@ func main() {
+		MistLoadBalancerTemplate: config.mistLoadBalancerTemplate,
+	})
+	go func() {
+		err := n.Balancer.Start()
+		if err != nil {
+			glog.Fatal(err)
 		}
 	}()
 
+	// Start HTTP servers
+	go n.startCatalystWebServer()
+	go n.startInternalWebServer()
+
 	go func() {
 		c := make(chan os.Signal, 1)
 		signal.Notify(c, syscall.SIGQUIT, syscall.SIGTERM, syscall.SIGINT)
 		for {
 			s := <-c
 			glog.Errorf("caught signal=%v killing MistUtilLoad", s)
-			killchan <- true
+			n.Balancer.Kill()
 		}
 	}()
 
-	// Everything past this is booting up Serf
-	tmpFile, err := writeSerfConfig(&serfConfig)
-	if err != nil {
-		glog.Fatalf("Error writing serf config: %s", err)
-	}
-	defer os.Remove(tmpFile)
-	glog.V(6).Infof("Wrote serf config to %s", tmpFile)
-
-	ui := &cli.BasicUi{Writer: os.Stdout}
-
-	// copied from:
-	// https://github.com/hashicorp/serf/blob/a2bba5676d6e37953715ea10e583843793a0c507/cmd/serf/commands.go#L20-L25
-	// we should consider invoking serf directly instead of wrapping their CLI helper
-	commands := map[string]cli.CommandFactory{
-		"agent": func() (cli.Command, error) {
-			a := &agent.Command{
-				Ui:         ui,
-				ShutdownCh: make(chan struct{}),
-			}
-			return a, nil
-		},
-	}
-
-	cli := &cli.CLI{
-		Args:     []string{"agent", "-config-file", tmpFile},
-		Commands: commands,
-		HelpFunc: cli.BasicHelpFunc("catalyst-node"),
-	}
-
-	exitCode, err := cli.Run()
-	if err != nil {
-		glog.Fatalf("Error executing CLI: %s\n", err.Error())
-		os.Exit(1)
-	}
-
-	os.Exit(exitCode)
-}
-
-func parseSerfConfig(config *agent.Config, retryJoin, serfTags *string) {
-	if *retryJoin != "" {
-		config.RetryJoin = strings.Split(*retryJoin, ",")
-	}
-	if *serfTags != "" {
-		if config.Tags == nil {
-			config.Tags = make(map[string]string)
-		}
-		for _, t := range strings.Split(*serfTags, ",") {
-			kv := strings.Split(t, "=")
-			if len(kv) == 2 {
-				k, v := kv[0], kv[1]
-				config.Tags[k] = v
-			} else {
-				glog.Fatalf("failed to parse serf tag, --serf-tag=k1=v1,k2=v2 format required: %s", t)
-			}
-		}
-	}
-}
-
-func writeSerfConfig(config *agent.Config) (string, error) {
-	// Two steps to properly serialize this as JSON: https://github.com/spf13/viper/issues/816#issuecomment-1149732004
-	items := map[string]interface{}{}
-	if err := mapstructure.Decode(config, &items); err != nil {
-		return "", err
-	}
-	b, err := json.Marshal(items)
-
-	if err != nil {
-		return "", err
-	}
-
-	// Everything after this is booting up serf with our provided config flags:
-	tmpFile, err := ioutil.TempFile(os.TempDir(), "serf-config-*.json")
-	if err != nil {
-		return "", err
-	}
-
-	// Example writing to the file
-	if _, err = tmpFile.Write(b); err != nil {
-		return "", err
-	}
-
-	// Close the file
-	if err := tmpFile.Close(); err != nil {
-		return "", err
+	// Start main reconcillation loop - get members from serf and update MistUtilLoad
+	for {
+		members := <-memberChan
+		n.Balancer.UpdateMembers(members)
```

Comment:
Not sure it's related to this PR, but we should start using `context` and `timeouts` for all network calls. E.g. here, if Mist does not reply or takes long to reply, it'll block the whole member update functionality. The same if when Serf does not reply.